spanner/api/Spanner.Samples/InsertDataAsync.cs

```diff
@@ -55,34 +56,40 @@ public async Task InsertDataAsync(string projectId, string instanceId, string da
             new Album { SingerId = 2, AlbumId = 2, AlbumTitle = "Forever Hold your Peace" },
             new Album { SingerId = 2, AlbumId = 3, AlbumTitle = "Terrified" },
         };
-        // Create connection to Cloud Spanner.
-        using var connection = new SpannerConnection(connectionString);
-        await connection.OpenAsync();
 
-        await Task.WhenAll(singers.Select(singer =>
+        using (TransactionScope scope = new TransactionScope(
```

Comment:
Just for the sake of figuring out what the error is, can you do a normal begin transaction from the connection, and set that transaction manually on all the commands? If that works, then the problem is in the ambient transaction mechanism.

Reply:
Ah, no, scratch that, the error is because when you get to execute the mutations that insert on the Album, the ones that inserted on the Singers table are not yet "effective".
So we are keeping the mutations in the order that we are receiving them, and in this code I'm pretty certain that all the Singers mutations will be received by the transaction before all the Albums mutations.
Can it be that in the backend Spanner does not necessarily respects that order?

(It was not failing before because each command was being individually executed).

I will try and debug this on Monday.

Reply:
I looked into this, and the problem is a race condition that occurs in the `VolatileResourceManager` implementation in the Spanner client, although I don't know whether that should be considered a bug in the `VolatileResourceManager` or wrong usage in this case (although my feeling says that this is a bug). I can reproduce this locally using a mock server and see exactly which requests and mutations are actually sent to Spanner in this use case.

What happens is the following:
1. The insertions of the `Singer` records are executed in parallel by [this call](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/blob/a6a3332cf32010bff2b3e281c6031e7d00ce0bca/spanner/api/Spanner.Samples/InsertDataAsync.cs#L67) to `Task.WhenAll`.
2. Each of these calls will get to [this point](https://github.com/googleapis/google-cloud-dotnet/blob/00dcfae3f83ab591dcc5d372c1d5dc17594d7653/apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/VolatileResourceManager.cs#L203) in `VolatileResourceManager` and see a `null` value for `_transaction`. Each of the calls will therefore also call `_spannerConnection.BeginTransactionAsync` and assign the returned value to `_transaction`. 
3. The last of the calls in step 2 to finish will eventually assign the returned transaction to `_transaction`, and that is the `SpannerTransaction` that will be used for the remainder of the ambient transaction. This will normally be the last mutation, i.e. the Singer with id 5.
4. The insertions of the Albums will use the transaction that was created by the last Singer mutation, and this will fail as it will only contain Singer 5 (and not Singer 1).
5. The insertions of Singers 1 to 4 will be added to other transactions that are never committed. The sessions that are used by these transactions are (AFAICT) leaked from the pool.

Assuming that executing multiple commands in parallel on the same (ambient) transaction is a valid use case, then we need some synchronization in `VolatileResourceManager` that will ensure that only one `SpannerTransaction` is ever created and assigned to it, and that it can only be committed once (and probably a couple of other things that I haven't noticed yet).
If executing multiple commands in parallel on the same ambient transaction is not a valid use case, then we should still have a closer look at this, as I think this example will cause 4 sessions to be leaked from the pool (the 4 sessions that are associated with the transactions that are created by the first 4 Singer records).

Reply:
Based on the internal discussion, I'm now using a `RetriableTransaction` so this is no longer an issue. This is also good in the sense that now we finally have a code sample that shows how to use a RetriableTransaction. Before, the only examples were in the snippets and tests in the main repo.